File: CardCreator.cs
// CardCreator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Processing;
using SixLabors.ImageSharp.Drawing.Processing;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.Fonts;

class CardCreator
{
    static Dictionary<int, (Color bg, Color text)> Templates = new()
    {
        {1, (Color.White, Color.Black)},
        {2, (Color.FromRgb(255,215,0), Color.FromRgb(139,0,0))},
        {3, (Color.FromRgb(173,216,230), Color.FromRgb(255,69,0))}
    };

    static Color HexToColor(string hex)
    {
        hex = hex.TrimStart('#');
        return Color.FromRgb(
            Convert.ToByte(hex.Substring(0,2), 16),
            Convert.ToByte(hex.Substring(2,2), 16),
            Convert.ToByte(hex.Substring(4,2), 16)
        );
    }

    static void CreateCard(string name, int? age, string message, int template,
                           string bgHex, string textHex, string photoPath,
                           string bgImagePath, string output)
    {
        int width = 800, height = 600;
        using var image = new Image<Rgba32>(width, height);

        // Background
        Color bgColor = bgHex != null ? HexToColor(bgHex) : Templates[template].bg;
        if (bgImagePath != null && File.Exists(bgImagePath))
        {
            using var bgImg = Image.Load(bgImagePath);
            bgImg.Mutate(x => x.Resize(width, height));
            image.Mutate(ctx => ctx.DrawImage(bgImg, new Point(0,0), 1f));
        }
        else
        {
            image.Mutate(ctx => ctx.Fill(bgColor));
        }

        // Text color
        Color textColor = textHex != null ? HexToColor(textHex) : Templates[template].text;

        // Load font (system fallback)
        var fontCollection = new FontCollection();
        // Try to load a system font, fallback to default
        FontFamily family;
        try {
            family = SystemFonts.Families.First(f => f.Name.Contains("Liberation") || f.Name.Contains("Arial"));
        } catch {
            family = SystemFonts.Families.First();
        }
        var fontBold = family.CreateFont(60, FontStyle.Bold);
        var fontRegular = family.CreateFont(50, FontStyle.Regular);
        var fontItalic = family.CreateFont(30, FontStyle.Italic);

        var center = new Point(width/2, 0);
        var options = new TextOptions(fontBold) { Origin = new PointF(width/2, 80), HorizontalAlignment = HorizontalAlignment.Center, VerticalAlignment = VerticalAlignment.Top };

        image.Mutate(ctx => {
            ctx.DrawText(options, "🎉 Happy Birthday! 🎉", textColor);

            options.Font = fontBold;
            options.Origin = new PointF(width/2, 200);
            ctx.DrawText(options, name, textColor);

            if (age.HasValue) {
                options.Font = fontRegular;
                options.Origin = new PointF(width/2, 280);
                ctx.DrawText(options, age.Value.ToString(), textColor);
            }

            // Wrap message
            var words = message.Split(' ');
            var lines = new List<string>();
            var line = "";
            foreach (var w in words) {
                if ((line + " " + w).Length <= 30) {
                    line = string.IsNullOrEmpty(line) ? w : line + " " + w;
                } else {
                    lines.Add(line);
                    line = w;
                }
            }
            if (!string.IsNullOrEmpty(line)) lines.Add(line);
            options.Font = fontItalic;
            int y = 380;
            foreach (var l in lines) {
                options.Origin = new PointF(width/2, y);
                ctx.DrawText(options, l, textColor);
                y += 40;
            }

            options.Font = fontRegular;
            options.Origin = new PointF(width/2, 520);
            ctx.DrawText(options, "🎂  🎈  🎁", textColor);

            // Photo
            if (photoPath != null && File.Exists(photoPath)) {
                using var photo = Image.Load(photoPath);
                int size = 150;
                photo.Mutate(x => x.Resize(size, size));
                ctx.DrawImage(photo, new Point(width - size - 20, 20), 1f);
            }
        });

        image.SaveAsPng(output);
        Console.WriteLine($"Card saved to {output}");
    }

    static void Main(string[] args)
    {
        var parsed = ParseArgs(args);
        if (!parsed.ContainsKey("n") || !parsed.ContainsKey("m")) {
            Console.Error.WriteLine("Error: -n and -m are required");
            return;
        }
        string name = parsed["n"];
        string message = parsed["m"];
        int? age = parsed.ContainsKey("a") ? int.Parse(parsed["a"]) : (int?)null;
        int template = parsed.ContainsKey("t") ? int.Parse(parsed["t"]) : 1;
        if (template < 1 || template > 3) {
            Console.Error.WriteLine("Template must be 1-3");
            return;
        }
        string bg = parsed.GetValueOrDefault("bg");
        string text = parsed.GetValueOrDefault("text");
        string photo = parsed.GetValueOrDefault("photo");
        string bgImage = parsed.GetValueOrDefault("bg-image");
        string output = parsed.GetValueOrDefault("o", "card.png");

        CreateCard(name, age, message, template, bg, text, photo, bgImage, output);

        if (parsed.ContainsKey("html")) {
            string htmlFile = parsed["html"];
            string bgColor = bg ?? "#FFFFFF";
            string textColor = text ?? "#000000";
            string ageHtml = age.HasValue ? $"<div class=\"age\">{age}</div>" : "";
            var html = $@"<!DOCTYPE html>
<html><head><title>Birthday Card</title>
<style>body{{font-family:sans-serif;text-align:center;background:#f0f0f0;}}
.card{{background:{bgColor};color:{textColor};border-radius:20px;padding:40px;max-width:600px;margin:50px auto;box-shadow:0 4px 8px rgba(0,0,0,0.2);}}
h1{{font-size:3em;}}.name{{font-size:2.5em;}}.age{{font-size:2em;}}.msg{{font-size:1.5em;}}</style>
</head><body><div class=""card"">
<h1>🎉 Happy Birthday! 🎉</h1>
<div class=""name"">{name}</div>
{ageHtml}
<div class=""msg"">{message}</div>
<div style=""font-size:3em;"">🎂 🎈 🎁</div>
</div></body></html>";
            File.WriteAllText(htmlFile, html);
            Console.WriteLine($"HTML card saved to {htmlFile}");
        }
    }

    static Dictionary<string, string> ParseArgs(string[] args)
    {
        var dict = new Dictionary<string, string>();
        for (int i = 0; i < args.Length; i++)
        {
            if (args[i].StartsWith("-"))
            {
                string key = args[i].TrimStart('-');
                if (i + 1 < args.Length && !args[i+1].StartsWith("-"))
                    dict[key] = args[++i];
                else
                    dict[key] = "";
            }
        }
        return dict;
    }
}

File: card_creator.py
# card_creator.py
import sys, os, argparse, json
from PIL import Image, ImageDraw, ImageFont, ImageFilter
import textwrap

TEMPLATES = {
    1: {"name": "Classic", "bg": (255,255,255), "text": (0,0,0), "font_size": 60},
    2: {"name": "Modern", "bg": (255,215,0), "text": (139,0,0), "font_size": 70},
    3: {"name": "Playful", "bg": (173,216,230), "text": (255,69,0), "font_size": 65}
}

def create_card(name, age, message, template, bg_color, text_color, photo_path, bg_image_path, output):
    # Use template defaults if colors not overridden
    if bg_color is None:
        bg_color = TEMPLATES[template]["bg"]
    else:
        bg_color = tuple(int(bg_color.lstrip('#')[i:i+2], 16) for i in (0,2,4))
    if text_color is None:
        text_color = TEMPLATES[template]["text"]
    else:
        text_color = tuple(int(text_color.lstrip('#')[i:i+2], 16) for i in (0,2,4))

    # Create base image
    width, height = 800, 600
    if bg_image_path and os.path.exists(bg_image_path):
        bg_img = Image.open(bg_image_path).resize((width, height))
        img = bg_img.convert("RGB")
    else:
        img = Image.new("RGB", (width, height), bg_color)

    draw = ImageDraw.Draw(img)
    try:
        font_large = ImageFont.truetype("/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf", 60)
        font_medium = ImageFont.truetype("/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf", 40)
        font_small = ImageFont.truetype("/usr/share/fonts/truetype/liberation/LiberationSans-Italic.ttf", 30)
    except:
        font_large = ImageFont.load_default()
        font_medium = ImageFont.load_default()
        font_small = ImageFont.load_default()

    # Draw title
    title = "🎉 Happy Birthday! 🎉"
    draw.text((width//2, 50), title, font=font_large, fill=text_color, anchor="mt")

    # Draw name
    draw.text((width//2, 200), name, font=font_large, fill=text_color, anchor="mt")

    # Draw age if provided
    if age:
        draw.text((width//2, 280), str(age), font=font_medium, fill=text_color, anchor="mt")

    # Draw message with wrapping
    wrapped = textwrap.fill(message, width=20)
    draw.text((width//2, 380), wrapped, font=font_small, fill=text_color, anchor="mt")

    # Draw emoji decorations (just text)
    draw.text((width//2, 500), "🎂  🎈  🎁", font=font_medium, fill=text_color, anchor="mt")

    # Embed photo if provided
    if photo_path and os.path.exists(photo_path):
        try:
            photo = Image.open(photo_path)
            # Resize to fit in top-right corner
            photo.thumbnail((150, 150))
            img.paste(photo, (width - photo.width - 20, 20))
        except Exception as e:
            print(f"Warning: could not embed photo: {e}", file=sys.stderr)

    # Apply a subtle filter for better look (optional)
    # img = img.filter(ImageFilter.SHARPEN)

    img.save(output)
    print(f"Card saved to {output}")

def main():
    parser = argparse.ArgumentParser(description="Birthday Card Creator")
    parser.add_argument('-n', '--name', required=True, help="Recipient's name")
    parser.add_argument('-a', '--age', type=int, help="Age (optional)")
    parser.add_argument('-m', '--message', required=True, help="Custom message")
    parser.add_argument('-t', '--template', type=int, choices=[1,2,3], default=1, help="Template (1-3)")
    parser.add_argument('--bg', help="Background color (hex, e.g. #FF0000)")
    parser.add_argument('--text', help="Text color (hex)")
    parser.add_argument('--photo', help="Path to photo to embed")
    parser.add_argument('--bg-image', help="Path to background image")
    parser.add_argument('-o', '--output', default="card.png", help="Output PNG file")
    parser.add_argument('--html', help="Export to HTML file (optional)")
    args = parser.parse_args()

    create_card(args.name, args.age, args.message, args.template, args.bg, args.text, args.photo, args.bg_image, args.output)

    if args.html:
        html_content = f"""<!DOCTYPE html>
<html><head><title>Birthday Card</title>
<style>body{{font-family:sans-serif;text-align:center;background:#f0f0f0;}}
.card{{background:{args.bg or '#FFFFFF'};color:{args.text or '#000000'};
border-radius:20px;padding:40px;max-width:600px;margin:50px auto;box-shadow:0 4px 8px rgba(0,0,0,0.2);}}
h1{{font-size:3em;}} .name{{font-size:2.5em;}} .age{{font-size:2em;}} .msg{{font-size:1.5em;}}</style>
</head><body><div class="card">
<h1>🎉 Happy Birthday! 🎉</h1>
<div class="name">{args.name}</div>
{'' if not args.age else f'<div class="age">{args.age}</div>'}
<div class="msg">{args.message}</div>
<div style="font-size:3em;">🎂 🎈 🎁</div>
</div></body></html>"""
        with open(args.html, 'w') as f:
            f.write(html_content)
        print(f"HTML card saved to {args.html}")

if __name__ == "__main__":
    main()
